lib/Conversion/PolyToStandard/PolyToStandard.cpp

```diff
@@ -60,11 +66,51 @@ struct ConvertAdd : public OpConversionPattern<AddOp> {
 
   using OpConversionPattern::OpConversionPattern;
 
+  // Convert add lowers a poly.add operation to arith operations. A poly.add
+  // operation is defined within the polynomial ring. Coefficients are added
+  // element-wise as elements of the ring, so they are performed modulo the
+  // coefficient modulus.
+  //
+  // Modular addition is performed by sign extending both arguments and
+  // performing (N+1)-bit addition. Then, if z = (x + y) is the (N+1)-bit sum,
```

Comment:
I can't help but wonder if it would make more sense to do this optimization in a separate pass/pattern to keep this lowering simpler. E.g,. could we later match on patterns that look like addui_extended followed by remui by a constant, and fold those greedily? Shouldn't it stand to reason that there are already such optimizations in MLIR? Or do we think that later passes/canonicalizations would muck that pattern up so it would be hard to match? I also wonder what existing canonicalizations might apply to a more straightforward lowering (addui + remui) vs this one---or maybe vice versa that eixsting canonicalizations would apply better to this form! I'm just unaware and would hence bias toward the simplest option.

In other words, I feel that if we can avoid mixing optimizations and conversions in the same pass, we should. Unless we have concrete evidence that this is the best place to do the optimization.

Reply:
> addui_extended

The problem is that `addui_extended` returns an overflow bit, not the result - so I don't really see how to get the remainder unless I applied that first. Perhaps I could do something clever by doing `a + (-mod + b)` when the overflow bit is 1. But that seems just the same as the optimization here to avoid remui.

> Shouldn't it stand to reason that there are already such optimizations in MLIR?

I would expect so only if they handled general modular arithmetic, but I'm not sure they do expect for power of 2 optimizations.

Reply:
> Unless we have concrete evidence that this is the best place to do the optimization.

I guess I can understand that - I thought that lowerings are generally where we would want to implement optimized versions, but perhaps the patterns that can be recognized should be optimized in a rewriter. I'm not super fussy on this, just wanted to get some implementation down to start with. The addui_extended route didn't seem to work, so I could naively do a remui and trunci, that works for me. Rewriting that pattern (we can do like "barret reduction rewrite" for muls too) works

Reply:
Cleared this up in a chat:

case 1: no assumptions, compute `c1, n1 = addui_extended` then `c0 = ((c1 % cmod) + (n1 2^N % cmod)) % cmod`
case 2: cmod divides 2^N, compute `c1 = addui` (allowing overflow) then `c0 = c1 % cmod`